react/OrderCoupon.tsx

```diff
@@ -57,37 +62,62 @@ export const OrderCouponProvider = compose(
     insertCoupon(coupon)
   }
 
-  const insertCoupon = (coupon: string) => {
-    return enqueue(async () => {
-      const mutationResult = await InsertCoupon({
-        variables: {
-          text: coupon,
-        },
+  const isQueueBusy = useRef(false)
+  useEffect(() => {
+    const unlisten = listen('Pending', () => (isQueueBusy.current = true))
+    return unlisten
+  })
+  useEffect(() => {
+    const unlisten = listen('Fulfilled', () => (isQueueBusy.current = false))
+    return unlisten
+  })
+
+  const insertCoupon = useCallback(
+    (coupon: string) => {
+      const marketingData = { ...orderForm.marketingData, coupon }
+
+      setOrderForm({
+        ...orderForm,
+        marketingData: marketingData,
       })
 
-      const newCoupon = mutationResult.data.insertCoupon.code
-        ? mutationResult.data.insertCoupon.code
-        : ''
+      setShowPromoButton(true)
 
-      if (newCoupon) {
-        setCoupon(newCoupon)
-        setIsShowingPromoButton(true)
-      } else {
-        setErrorKey(CODE_DOESNT_EXIST)
+      const task = async () => {
+        const {
+          data: { insertCoupon: newOrderForm },
+        } = await InsertCoupon({
+          variables: {
+            text: coupon,
+          },
+        })
+
+        return newOrderForm
       }
-    })
-  }
+
+      enqueue(task)
```

Comment:
Since we only accept one coupon at a time, you can pass a [second argument](https://github.com/vtex-apps/order-manager#enqueuetask---promise-id-string-promise) to `enqueue` to cancel a coupon task if a newer one is pushed to the queue.